### PR TITLE
shell: capture job shell error messages in designated output file

### PR DESCRIPTION
**Pull request by grondo**
This PR is a stopgap fix for #4117. 

When an output or error file is specified, re-route log messages to the same output file as stderr, instead of sending them to the output eventlog.

Since the output files aren't opened until after the shell has connected to Flux and initialized, some log messages and errors still end up in the output eventlog, e.g. if something goes wrong before shell.init, but at least there isn't likely to be an _empty_ output file when a job fails after it gets that far, so the spirit of the issue is resolved.

Eventually, perhaps a better fix would be to have the exec system read the output and error options from the jobspec and, if the output type is `file`, pass this along in new `--output` and `--error` options to the job shell. This would allow the job shell to open output file immediately, and even the earliest of errors could be captured. However, I'm not sure at this point if that is the "right" fix.

before:
```console
$ id=$(flux mini submit --wait --output=example.out nosuchcommand) && flux job status $id; ls -l example.out; cat example.out
ƒJSBRFaBzRm: exception: type=exec note=task 0: start failed: nosuchcommand: No such file or directory
-rw-r--r-- 1 grondo grondo 0 Feb 11 07:43 example.out
``` 

after:
```console
$ id=$(flux mini submit --wait --output=example.out nosuchcommand) && flux job status $id; ls -l example.out; cat example.out
ƒ2uW3UFh: exception: type=exec note=task 0: start failed: nosuchcommand: No such file or directory
-rw-r--r-- 1 grondo grondo 85 Feb 11 07:44 example.out
flux-shell[0]: FATAL: task 0: start failed: nosuchcommand: No such file or directory
```

### CODE REVIEW

File: src/shell/output.c [hunk 1]
```diff
@@ -409,15 +409,17 @@ static int shell_output_file (struct shell_output *out)
...
                                       json_t *o,
                                       flux_msg_handler_t *mh) // may be NULL
 {
     bool eof = false;
     json_t *entry;
 
-    if (iodecode (o, NULL, NULL, NULL, NULL, &eof) < 0)
+    if (strcmp (type, "data") == 0
+        && iodecode (o, NULL, NULL, NULL, NULL, &eof) < 0)
         goto error;
```
Comment (garlick) [inline src/shell/output.c]: errno is not set if the strcmp condition is true but `shell_output_write_leader() ` is expected to set errno or `flux_respond_error()` will fail with EINVAL.

File: src/shell/output.c [hunk 2]
```diff
@@ -409,15 +409,17 @@ static int shell_output_file (struct shell_output *out)
 }
```
Comment (garlick) [inline src/shell/output.c:453]: in commit message for 8e13b6ee8e057b2c0a879672e5d357f5867ce76b s/feild/field/ and s/commont/common/

File: src/shell/output.c [hunk 3]
```diff
@@ -345,6 +345,47 @@ static int shell_output_write_fd (int fd, const void *buf, size_t len)
...
+        output_type = out->stderr_type;
+        ofp = &out->stderr_file;
+    }
+    if ((output_type == FLUX_OUTPUT_TYPE_FILE) && len > 0) {
+        if (ofp->label
+            && shell_output_label (ofp, rank) < 0)
+            return -1;
+        if (shell_output_write_fd (ofp->fdp->fd, data, len) < 0)
+            return -1;
+    }
```
Comment (garlick) [inline src/shell/output.c:384]: Is `data` leaked on error?

File: src/shell/output.c [hunk 4]
```diff
@@ -84,7 +84,6 @@ struct shell_output {
     struct eventlogger *ev;
```
Comment (garlick) [inline src/shell/output.c:84]: In commit message for 31277bad4117442931d3adcf0e7a6a1981751990 s/refernence/reference/

**Review by garlick — COMMENTED** [review]
Whoops I started making review comments but I think you fixed some of them as I was making them.  I'll submit what I have so far in case anything is useful and then try again when you're ready.

File: src/shell/output.c (hunk 1 above)
Comment (garlick) [reply src/shell/output.c]: Pretty sure you just made this a non-error.

Comment (grondo) [reply src/shell/output.c]: Yeah, iirc we don't need to separately check for `eof` anymore so this conditional could be dropped.

File: src/shell/output.c (hunk 3 above)
Comment (grondo) [reply src/shell/output.c:384]: yes, good catch!

Comment (grondo) [reply src/shell/output.c]: I was going to add a separate commit for this fix since it was present in existing code, however I just squashed it into the refactor commit. Hope that was ok!

File: src/shell/output.c [hunk 5]
```diff
@@ -458,7 +459,7 @@ static void shell_output_decref (struct shell_output *out,
                                  flux_msg_handler_t *mh)
 {
     if (--out->refcount == 0) {
-        flux_msg_handler_destroy (mh);
+        flux_msg_handler_stop (mh);
```
Comment (chu11) [inline src/shell/output.c:462]: i think this flux_msg_handler_destroy() -> flux_msg_handler_stop() is in the wrong commit?

Comment (grondo) [reply src/shell/output.c:462]: Good catch, it should have never been destroyed and I tried to fix up a development typo in a rebase and must have messed that up. Let me see if I can fix it up. Thanks!

Comment (grondo) [reply src/shell/output.c:462]: Ah, yeah it was supposed to be squashed into 4d1e7f5

Comment (grondo) [reply src/shell/output.c:462]: Ok, I think I've fixed that.

**Comment by chu11** [comment]
re-pushed, since the last time I squash

- add `sdexec_cancel()`, to handle corner case when `sdprocess_exec()` fails.
- re-add stdout/stderr logging to `exec.eventlog`, make logging configurable and fix bug in which logging didn't work
- did some refactoring
- add tests for much fo the above
- added basic system / multiuser tests for `sdexec`

Note that the system / multiuser tests use `system.exec.sd.test=true` attribute to make sure tests are run under sdexec / systemd.  This can be altered later if we make `sdexec` the default.  Unsure if we'd want to update the default `exec.toml` or create one in the fluxorama image that makes `sdexec` the default.   Thoughts?

**Comment by grondo** [comment]
rebased.

**Review by garlick — APPROVED** [review]
This LGTM!

**Comment by grondo** [comment]
Thanks! Set MWP
